fitbert/fitb.py

```diff
@@ -35,90 +40,196 @@ def softmax(x):
-        but modified for list of words
-
-        also useful,
-        https://stackoverflow.com/questions/54978443/predicting-missing-words-in-a-sentence-natural-language-processing-model
-
-        @TODO this is not optimized! It loops instead of batching, forgive me, Tensor Gods!
-        """
-        option_2_token_count: Dict[str, int] = defaultdict(int)
-        for word in words:
-            toks = self.tokenizer.tokenize(word)
-            option_2_token_count[word] += len(toks)
-
-        token_count_2_option: Dict[int, List[str]] = defaultdict(list)
-        for k, v in option_2_token_count.items():
-            token_count_2_option[v].append(k)
-
-        final_scores: Dict[str, List[float]] = {}
-        for tok_len, options_of_given_length in token_count_2_option.items():
-
-            pre, target_s, post = sent.split("***")
-            if "mask" in target_s.lower():
-                target = ["[MASK]"] * tok_len
-            else:
-                # not sure how to get here, but it was in the original
-                target = self.tokenizer.tokenize(target_s)
-
-            tokens = ["[CLS]"] + self.tokenizer.tokenize(pre)
-            target_idx = len(tokens)
-            tokens += target + self.tokenizer.tokenize(post) + ["[SEP]"]
-            input_ids = self.tokenizer.convert_tokens_to_ids(tokens)
-            for option in options_of_given_length:
-                try:
-                    toks = self.tokenizer.tokenize(option)
-                    tok_ids = self.tokenizer.convert_tokens_to_ids(toks)
-                except KeyError:
-                    print("couldn't convert tokens to IDs, ", toks)
-                    return None
-                tens = torch.LongTensor(input_ids).unsqueeze(0)
-
-                # don't need gradients at inference time, thank god
-                # speeds things up considerably
-                with torch.no_grad():
-                    pred = self.bert(tens)
-                scores = []
-                for i in range(tok_len):
-                    res = pred[0, target_idx + i]
-                    prob = self.softmax(res)
-                    score = float(prob[tok_ids[i]].item())
-                    # to explain the res[tok_ids] syntax:
-                    # >>> a = torch.tensor([1.0, 1.1, 1.2, 1.3])
-                    # >>> a[[1,2]]
-                    # tensor([1.1000, 1.2000])
-                    scores.append(score)
-                final_scores[option] = reduce(operator.mul, scores, 1)
-        return final_scores
+    def _tokens_to_masked_ids(self, tokens, mask_ind):
+        masked_tokens = tokens[:]
+        masked_tokens[mask_ind] = "[MASK]"
+        masked_tokens = ["[CLS]"] + masked_tokens + ["[SEP]"]
+        masked_ids = self.tokenizer.convert_tokens_to_ids(masked_tokens)
+        return masked_ids
+
+    def _get_sentence_probability(self, sent: str) -> float:
+
+        tokens = self.tokenizer.tokenize(sent)
+        input_ids = (
```

Comment:
I'm trying to learn `seq`... does wrapping it with parens make it a `tuple`? something else?

Reply:
Hi, from pyfunctional docs: 

`seq = Stream()`

where: 

Stream is a class with `__call__` so we can use syntax `seq(...)`:

        Create a Sequence using a sequential ExecutionEngine.
        If args has more than one argument then the argument list becomes the sequence.
        If args[0] is primitive, a sequence wrapping it is created.
        If args[0] is a list, tuple, iterable, or Sequence it is wrapped as a Sequence.